features/include/pcl/features/impl/gasd.hpp

```diff
@@ -0,0 +1,414 @@
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::computeFeature (PointCloudOut &output)
+  {
+    // compute alignment transform using reference frame
+    computeAlignmentTransform ();
+
+    // align point cloud
+    pcl::transformPointCloud (*surface_, *indices_, shape_samples_, transform_);
+
+    const size_t shape_grid_size = shape_half_grid_size_ * 2;
+
+    // each histogram dimension has 2 additional bins, 1 in each boundary, for performing interpolation
+    std::vector < Eigen::VectorXf
+        > shape_hists ( (shape_grid_size + 2) * (shape_grid_size + 2) * (shape_grid_size + 2), Eigen::VectorXf::Zero (shape_hists_size_ + 2));
+
+    Eigen::Vector4f centroid_p = Eigen::Vector4f::Zero ();
+
+    // compute normalization factor for distances between samples and centroid
+    Eigen::Vector4f far_pt;
+    pcl::getMaxDistance (shape_samples_, centroid_p, far_pt);
+    far_pt[3] = 0;
+    const float distance_normalization_factor = (centroid_p - far_pt).norm ();
+
+    // compute normalization factor with respect to axis-aligned bounding cube centered on the origin
+    Eigen::Vector4f min_pt, max_pt;
+    pcl::getMinMax3D (shape_samples_, min_pt, max_pt);
+
+    max_coord_ = std::max (min_pt.head<3> ().cwiseAbs ().maxCoeff (), max_pt.head<3> ().cwiseAbs ().maxCoeff ());
+
+    // normalize sample contribution with respect to the total number of points in the cloud
+    hist_incr_ = 100.0f / static_cast<float> (shape_samples_.size () - 1);
+
+    // for each sample
+    for (size_t i = 0; i < shape_samples_.size (); ++i)
+    {
+      // compute shape histogram array coord based on distance between sample and centroid
+      const Eigen::Vector4f p (shape_samples_[i].x, shape_samples_[i].y, shape_samples_[i].z, 0.0f);
+      const float d = p.norm ();
+
+      const float shape_grid_step = distance_normalization_factor / shape_half_grid_size_;
+
+      const float dist_hist_start = ((int) (d / shape_grid_step)) * shape_grid_step;
+
+      const float dist_hist_val = (d - dist_hist_start) / shape_grid_step;
+
+      const float dbin = dist_hist_val * shape_hists_size_;
+
+      // add sample to shape histograms, optionally performing interpolation
+      addSampleToHistograms (p, max_coord_, shape_half_grid_size_, shape_interp_, dbin, hist_incr_, shape_hists);
+    }
+
+    pos_ = 0;
+
+    // copy shape histograms to output
+    copyShapeHistogramsToOutput (shape_grid_size, shape_hists_size_, shape_hists, output, pos_);
+
+    // set remaining values of the descriptor to zero (if any)
+    std::fill (output.points[0].histogram + pos_, output.points[0].histogram + output.points[0].descriptorSize (), 0.0f);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT>
+  template <typename T> void
```

Comment:
Isn't type T equal to type PointInT? If so, you don't need the extra templates type.